### ISSUE TITLE
Change the driver to Npcap

### ISSUE THREAD

**Issue by CapitanShinChan** [google/gopacket#500]
Hi!

For Windows builds, the current driver and library used is WinPcap, but there is a new more resilient and with more features called Npcap (developed by the community around Nmap). I've been using it with gopacket for some weeks without any problem.

The comparison of features:
[Npcap vs Winpcap](https://nmap.org/npcap/vs-winpcap.html)

Also, Npcap was designed with the intention of replacing Winpcap, so it's almost 100% compatible with the 

More info here: [Npcap description](https://nmap.org/npcap/)

**Comment by Seumi** [comment]
I am also thinking about it! Npcap is more advanced than winpcap.
Could you tell me how you get gopacket work with ncap?
thanks a lot.

**Comment by CapitanShinChan** [comment]
Hi!

I followed this guide: https://stackoverflow.com/questions/38047858/compile-gopacket-on-windows-64bit

And then you just have to modify the source code of gopacket to use Npcap.dll instead of Winpcap.dll

**Comment by notti** [comment]
This should work with maybe two small things:
L63 -  L78 might also need removing thats:
```

#ifdef WIN32
#define pcap_statustostr pcap_strerror
// WinPcap also doesn't export pcap_can_set_rfmon and pcap_set_rfmon,
// as those are handled by separate libraries (airpcap).
// https://www.winpcap.org/docs/docs_412/html/group__wpcapfunc.html
// Stub out those functions here, returning values that indicate rfmon
// setting is unavailable/unsuccessful.
int pcap_can_set_rfmon(pcap_t *p) {
	return 0;
}
int pcap_set_rfmon(pcap_t *p, int rfmon) {
	return PCAP_ERROR;
}
#endif 
```
since those functions are present in npcap.

Additionally, if you have both npcap and winpcap on your computer installed, you might run into the trouble, that gopacket loads the winpcap version (since it lives in system32, which takes precedence over npcap in system32/npcap). Sadly I couldn't get the delay-loaded version working, where it would be possible to select this at runtime.

I'm currently working on a solution that works both with winpcap and npcap (including runtime selection between the both, so one can compile with either version - or eventually without having any sdk installed) - but this will take a bit of time.

**Comment by CapitanShinChan** [comment]
Thanks for sharing that @notti !

Probably you already checked this, but just in case:
https://nmap.org/npcap/guide/npcap-devguide.html#npcap-devguide-updating

This part specifically:
> You need to do the following one step:
> 
> Before calling LoadLibrary to load wpcap.dll, call SetDllDirectory to add C:\Windows\System32\Npcap\ to DLL search path.

**Comment by notti** [comment]
Yes, that was basically my starting point.
Sadly I couldn't get implicit linking with delay loading and path changing to work with golang (cgo doesn't seem to support the delay loading), so I'll want to give explicit linking a try and see if this can be integrated without disturbing the other OSes - and if possible without using cgo (should simplify compiling on windows a lot)

**Comment by gconnell** [comment]
My quick 2 cents:  I'd rather add support for npcap than replace winpcap with npcap.  In short, it would be nice if GoPacket worked for whatever type of underlying packet capture mechanism the user chose (winpcap, npcap, af_packet, pf_ring, libpcap, etc).  Also, straight-up switching off of winpcap could break existing users of that packet source.

**Comment by notti** [comment]
Yes, that's my plan - I don't want to break existing implementations. Especially since the licensing of npcap could pose some serious problems.

What I did so far is:

1. try out a small poc that can use either winpcap or npcap with implicit linking without using cgo (cgo is complicated to setup on windows and there are no standardpaths to put stuff..) -> https://github.com/notti/npcap_test/blob/master/main.go
2. Factor out all the C. stuff from pcap.go and move it to pcap_unix.go by introducing an internal api that pcap.go calls
3. Implement the new internal api in pcap_windows.go
4. Have a look at performance

So from the outside pcap.go still looks and works the same. Internally, on unix, it now calls into pcap_unix.go which does the stuff pcap.go did before. And on windows it now calls into pcap_windows.go which calls into wpcap.dll from either winpcap or npcap via the syscall package.

Basically, on unix it works the same. On windows it works with either winpcap or npcap (whichever is there, preferring npcap) and for building neither cgo nor winpcap/npcap sdks are needed.

So far I managed to do 1., 2. and 1/4 of 3. (omg pcap.go uses tons of pcap_-calls...)

**Comment by notti** [comment]
@CapitanShinChan, @Seumi  could you try out #568 and see if this one works for you?

**Comment by negbie** [comment]
@notti just a experience report on windows 10 64 bit. So far no problems. Building without npcap/winpcap sdks works. Tested it first with WinPcap 4.1.3. Installed it and checked the interfaces:

(Realtek PCIe GBE Family Controller)
(Microsoft)
(Microsoft)
(Microsoft)

Next I installed npcap-0.99-r7 without WinPcap API-compatible Mode so WinPcap wouldn't be uninstalled. Check the interfaces and I saw some additonal ones so I knew npcap was used over WinPcap.

(NdisWan Adapter)
(NdisWan Adapter)
(MS NDIS 6.0 LoopBack Driver)
(NdisWan Adapter)

Uninstalled WinPcap to make sure everything still works. Great job! Thank you!

**Comment by joe-conigliaro** [comment]
Also npcap has an install option to install in winpcap compatibility mode

**Comment by joe-conigliaro** [comment]
Would be nice to have the monitor mode on windows enabled

**Comment by brydavis** [comment]
@joe-conigliaro is right.
Installing [Npcap](https://nmap.org/npcap/) in "Wpcap API Compatibility Mode" did the trick.
